src/main/scala/com/imaginea/Main.scala

```diff
@@ -2783,6 +2867,88 @@ object Main extends App {
+      val cmd = command.trim
+      val cmdName = cmd.split("\\s+")(0)
+      val argsLine = cmd.substring(cmdName.length).trim
+      logger.info(s"Command to be executed [$cmdName] with args and options: $argsLine")
+      val pattern = "[^\\s\"']+|\"[^\"]*\"|'[^']*'".r
+      val regexMatcher = pattern.findAllMatchIn(argsLine)
+      val matchList = regexMatcher.map(m => m.group(0)).toList
+      map + ((cmdName, matchList))
+    }
+    commandsMap
+  }
+
+  def executePipedCommand(session: TerminalSession, commadLine: String): CommandResult = {
+    val commandsMap = parseCommandLine(commadLine)
+    val executionContext = session.currentCmdExecContext
+    val (result, currentContext) = commandsMap.foldLeft(List.empty[Line], executionContext) { (inputAndContext, map) =>
+      val (input, context) = inputAndContext
+      val (commandName, argsList) = map
+      val commandResult = commandName match {
+        case Constants.LIST_COMMAND => Command.executeListCommand(context, input, argsList)
+        case Constants.CD_COMMAND => throw new RuntimeException(s"Functionality yet to be implemented for [$commandName]")
+        case Constants.GREP_COMMAND => throw new RuntimeException(s"Functionality yet to be implemented for [$commandName]")
+        case _ => throw new RuntimeException(s"Unsupported command with name [$commandName]")
+      }
+      commandResult.currentContext match {
+        case Some(execContext) => (commandResult.result, execContext)
+        case None => throw new Exception(s"did not get execution context after executing Command with name [$commandName]")
+      }
+    }
+    CommandResult(result, Some(currentContext))
+  }
+
+  def executeSSHCommand(session: TerminalSession, command: String): CommandResult = {
+    val sshSessions = session.sshSessions
+    val result = sshSessions.map { sshSession =>
+      val mayBeSession = sshSession.session
+      Future {
+        val mayBeOutput = mayBeSession.map { session =>
+          val channel = session.openChannel("exec").asInstanceOf[ChannelExec]
+          channel.setPty(true)
+          channel.setCommand(command)
+          getOutputFromChannel(channel)
+        }
+        mayBeOutput match {
+          case Some(output) => Line(List(output), LineType.toLineType("RESULT"), sshSession.serverIp)
+          case None => throw new Exception("Session not found")
+        }
+      }
+    }
+    val futureList = Future.sequence(result)
+    val lines = Await.result(futureList, Duration(Long.MaxValue, TimeUnit.NANOSECONDS))
+    CommandResult(lines, None)
+  }
+
+  def getOutputFromChannel(channel: ChannelExec): String = {
+    val results = new StringBuilder
+    val sleepTimeInMilliSeconds = 1000L
+    val inputStream = channel.getInputStream
+    try {
+      channel.connect()
+      val buffer = new Array[Byte](1024 * 4)
+      // read output till the channel is closed with polling
+      while (channel.isConnected) {
```

Comment:
Saurabh, I have removed Thread.sleep and modified the code. Please take a look and let me know if its alright.